### PR TITLE
Avoid allocating strings in DefaultViewLocationCache

### LINKED ISSUE #3135: [PERF] Avoid allocating strings in DefaultViewLocationCache
As part of looking up a view, we construct a string based cache key in `DefaultViewLocationCache`. This results in allocation about 1.5Mb (~1% of all object allocations) across 3000 requests

defaultviewlocationcache-before

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #3135

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @rynowak

**Comment by pranavkm** [comment]
defaultviewlocationcache-after

**Comment by rynowak** [comment]
prkrishnPrkrishn/3135 is probably the most important bug we've encountered. Glad to see a fix in flight!

**Comment by pranavkm** [comment]
Better?

File: src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs [hunk 1]
```diff
@@ -45,46 +43,123 @@ public ViewLocationCacheResult Get([NotNull] ViewLocationExpanderContext context
...
             [NotNull] ViewLocationCacheResult value)
         {
-            var cacheKey = GenerateKey(context);
+            var cacheKey = GenerateKey(context, copyViewExpanderValues: true);
             _cache.TryAdd(cacheKey, value);
         }
 
-        internal static string GenerateKey(ViewLocationExpanderContext context)
+        // Internal for unit testing
+        internal static ViewLocationCacheKey GenerateKey(
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs]: Document the optimization behind `copyViewExpanderValues`

File: src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs [hunk 2]
```diff
@@ -45,46 +43,123 @@ public ViewLocationCacheResult Get([NotNull] ViewLocationExpanderContext context
...
+                    !string.Equals(x.AreaName, y.AreaName, StringComparison.Ordinal))
                 {
-                    keyBuilder.Append(CacheKeySeparator)
-                              .Append(item.Key)
-                              .Append(CacheKeySeparator)
-                              .Append(item.Value);
+                    return false;
                 }
+
+                if (x.Values == y.Values)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs]: Using `object.ReferenceEquals` if that's what you intend.

File: src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs [hunk 3]
```diff
@@ -45,46 +43,123 @@ public ViewLocationCacheResult Get([NotNull] ViewLocationExpanderContext context
...
+
+        // Internal for unit testing
+        internal struct ViewLocationCacheKey
+        {
+            public ViewLocationCacheKey(
+                string viewName,
+                string controllerName,
+                string areaName,
+                bool isPartial,
+                IDictionary<string, string> values)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs]: Is it possible for this to be `Dictionary`? or do you have to handle the `I`?

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs (hunk 3 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs]: The type that is passed via `ViewLocationExpanderContext` is an `IDictionary`.

File: src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs]: Does it make a difference for `IDictionary`s?

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultViewLocationCache.cs]: It's clearer what you intended to happen

**Pull request closed without merging**